test_vm/tests/extend_sectors_test.rs

```diff
@@ -0,0 +1,326 @@
+use fil_actors_runtime::cbor::serialize;
+use fil_actors_runtime::runtime::Policy;
+use fil_actors_runtime::test_utils::make_sealed_cid;
+use fil_actors_runtime::{
+    DealWeight, CRON_ACTOR_ADDR, EPOCHS_IN_DAY, STORAGE_MARKET_ACTOR_ADDR,
+    STORAGE_POWER_ACTOR_ADDR, SYSTEM_ACTOR_ADDR, VERIFIED_REGISTRY_ACTOR_ADDR,
+};
+use fvm_ipld_bitfield::BitField;
+use fvm_ipld_blockstore::MemoryBlockstore;
+use fvm_ipld_encoding::RawBytes;
+use fvm_shared::bigint::Zero;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::piece::PaddedPieceSize;
+use fvm_shared::sector::{RegisteredSealProof, SectorNumber, StoragePower};
+use test_vm::util::{
+    add_verifier, advance_by_deadline_to_epoch, advance_by_deadline_to_epoch_while_proving,
+    advance_by_deadline_to_index, advance_to_proving_deadline, apply_ok, create_accounts,
+    create_miner, publish_deal, submit_windowed_post,
+};
+use test_vm::{ExpectInvocation, VM};
+
+#[test]
+fn extend_sector_with_deals() {
+    let store = MemoryBlockstore::new();
+    let mut v = VM::new_with_singletons(&store);
+    let addrs = create_accounts(&v, 3, TokenAmount::from(10_000e18 as i128));
+    let seal_proof = RegisteredSealProof::StackedDRG32GiBV1P1;
+    let (owner, worker, verifier, verified_client) = (addrs[0], addrs[0], addrs[1], addrs[2]);
+    let sector_number: SectorNumber = 100;
+    let sealed_cid = make_sealed_cid(b"s100");
+    let policy = Policy::default();
+
+    // create miner
+
+    let miner_id = create_miner(
+        &mut v,
+        owner,
+        worker,
+        seal_proof.registered_window_post_proof().unwrap(),
+        TokenAmount::from(1_000e18 as i128),
+    )
+    .0;
+    let mut v = v.with_epoch(200);
+
+    //
+    // publish verified deals
+    //
+
+    // register verifier then verified client
+
+    add_verifier(&v, verifier, StoragePower::from((32_u64 << 40) as u128));
+    let add_client_params = AddVerifierClientParams {
+        address: verified_client,
+        allowance: StoragePower::from((32_u64 << 40) as u64),
+    };
+    apply_ok(
+        &v,
+        verifier,
+        *VERIFIED_REGISTRY_ACTOR_ADDR,
+        TokenAmount::zero(),
+        VerifregMethod::AddVerifiedClient as u64,
+        add_client_params,
+    );
+
+    // add market collateral for clients and miner
+    let mut collateral = TokenAmount::from(3e18 as i128);
+    apply_ok(
+        &v,
+        verified_client,
+        *STORAGE_MARKET_ACTOR_ADDR,
+        collateral.clone(),
+        MarketMethod::AddBalance as u64,
+        verified_client,
+    );
+    collateral = TokenAmount::from(64e18 as i128);
+    apply_ok(
+        &v,
+        worker,
+        *STORAGE_MARKET_ACTOR_ADDR,
+        collateral,
+        MarketMethod::AddBalance as u64,
+        miner_id,
+    );
+
+    // create 1 verified deal for total sector capacity for 6 months
+    let deal_start =
+        v.get_epoch() + max_prove_commit_duration(&Policy::default(), seal_proof).unwrap();
```

Comment:
I tried that and then had a hard time reasoning about the math around deal weights. The sector's lifetime, measured from its _activation_ has to be at least 6 months, which means the sector would have to expire after the deal if we wanted the deal to start earlier. I'm not sure whether having the sector expire later than the deal will work for the math in the test.

Reply:
Yeah you're right.  I was forgetting about precommit expiration validation.  It validates with the latest possible activation assuming the precommit sits for 1 month before activation.  This adds ~ a month to the minimum sector expiration for sectors being prove committed as soon as possible because you can't add a sector that expires sooner than 6 months after a point 1 month in the future i.e. you can't add a sector that expires sooner than 7 months.

You could keep the math the same by using an expiration of ~ 7 months and then increasing the expiration by this interval.  However then you would need to run through a longer expiration time when processing the sector's lifetime in the test.  So leaving it the way it is looks best for minimizing epochs.

--tangent-- we are not actually getting max 540 day limit wrong since we check that against precommit epoch, but if the max sector lifetime parameter (currently five years) was smaller than the max extension we would be lenient by an extra month.